src/main/java/io/confluent/copycat/hdfs/WAL.java

```diff
@@ -0,0 +1,161 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URI;
+
+import io.confluent.copycat.hdfs.WALFile.Reader;
+import io.confluent.copycat.hdfs.WALFile.Writer;
+
+public class WAL {
+
+  private static final Logger log = LoggerFactory.getLogger(WAL.class);
+  private static final String leaseException =
+      "org.apache.hadoop.hdfs.protocol.AlreadyBeingCreatedException";
+  private static final long MAX_SLEEP_INERVAL_MILLIS = 16000L;
+  private static final int MAX_NUM_RETRY = 5;
+
+  private WALFile.Writer writer = null;
+  private WALFile.Reader reader = null;
+  private FileSystem fs = null;
+  private String logFile = null;
+  private Configuration conf = null;
+  private long sleepIntervalMillis = 1000L;
+  private int num_retry = 0;
+
+  public WAL(Configuration conf, String url, String topicsDir, TopicPartition topicPart)
+      throws CopycatException {
+    try {
+      this.conf = conf;
+      fs = FileSystem.newInstance(URI.create(url), conf);
+      logFile = FileUtils.logFileName(url, topicsDir, topicPart);
+    } catch (IOException e) {
+      throw new CopycatException(e);
+    }
+  }
+
+  public void append(String tempFile, String committedFile) throws CopycatException {
+    try {
+      acquireLease();
+      WALEntry key = new WALEntry(tempFile);
+      WALEntry value = new WALEntry(committedFile);
+      writer.append(key, value);
+      writer.hsync();
+    } catch (IOException e) {
+      throw new CopycatException(e);
+    }
+  }
+
+  private void acquireLease() throws CopycatException {
+    long initialSleepIntervalMillis = sleepIntervalMillis;
+    while (sleepIntervalMillis < MAX_SLEEP_INERVAL_MILLIS) {
+      try {
+        if (writer == null) {
+          writer = WALFile.createWriter(conf,
+                                        Writer.file(new Path(logFile)),
+                                        Writer.appendIfExists(true));
+        }
+        log.info("Successfully acquired lease for {}", logFile);
+        break;
+      } catch (RemoteException e) {
+        if (e.getClassName().equals(leaseException)) {
+          sleepIntervalMillis = sleepIntervalMillis * 2;
+          log.info("Cannot acquire lease on WAL {}", logFile);
+          try {
+            Thread.sleep(sleepIntervalMillis);
+          } catch (InterruptedException ie) {
+            throw new CopycatException(ie);
+          }
+        } else {
+          throw new CopycatException(e);
+        }
+      } catch (IOException e) {
+        throw new CopycatException("Error creating writer for log file " + logFile, e);
+      }
+    }
+    if (sleepIntervalMillis >= MAX_SLEEP_INERVAL_MILLIS) {
+      if (num_retry <= MAX_NUM_RETRY) {
+        num_retry++;
+        throw new TransientException("Cannot acquire lease after timeout, will retry.");
+      } else {
+        throw new CopycatException("Cannot acquire lease after retry.");
+      }
+    }
+    sleepIntervalMillis = initialSleepIntervalMillis;
```

Comment:
I don't think this is actually reset properly. If you throw a `TransientException` above, it'll still be at the last value from the previous iteration when this method is called again.

It looks like `sleepIntervalMillis` isn't even used outside this method. Could we change this so we only have a constant like `INITIAL_SLEEP_INTERVAL_MILLIS = 1000L` and `sleepIntervalMillis` is made local to `acquireLease`?